tests/53groups/03local.pl

```diff
@@ -0,0 +1,199 @@
+
+   do => sub {
+      my ( $creator, $user ) = @_;
+
+      my $group_id;
+
+      matrix_create_group( $creator )
+      ->then( sub {
+         ( $group_id ) = @_;
+
+         matrix_invite_group_users( $creator, $group_id, $user );
+      })->then( sub {
+         matrix_accept_group_invite( $group_id, $user );
+      })->then( sub {
+         matrix_get_joined_groups( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_keys( $body, qw( groups ) );
+         assert_json_list( my $group_ids = $body->{groups} );
+
+         assert_deeply_eq( $group_ids, [ $group_id ] );
+
+         Future->done( 1 );
+      });
+   };
+
+test "Remove self from local group",
+   requires => [ local_admin_fixture( with_events => 0 ), local_user_fixture( with_events => 0 ) ],
+
+   do => sub {
+      my ( $creator, $user ) = @_;
+
+      my $group_id;
+
+      matrix_create_group( $creator )
+      ->then( sub {
+         ( $group_id ) = @_;
+
+         matrix_invite_group_users( $creator, $group_id, $user );
+      })->then( sub {
+         matrix_accept_group_invite( $group_id, $user );
+      })->then( sub {
+         matrix_get_joined_groups( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_list( my $group_ids = $body->{groups} );
+         assert_deeply_eq( $group_ids, [ $group_id ] );
+
+         matrix_remove_group_self( $user, $group_id );
+      })->then( sub {
+         matrix_get_joined_groups( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_list( my $group_ids = $body->{groups} );
+         assert_deeply_eq( $group_ids, [] );
+
+         Future->done( 1 );
+      });
+   };
+
+test "Remove other from local group",
+   requires => [ local_admin_fixture( with_events => 0 ), local_user_fixture( with_events => 0 ) ],
+
+   do => sub {
+      my ( $creator, $user ) = @_;
+
+      my $group_id;
+
+      matrix_create_group( $creator )
+      ->then( sub {
+         ( $group_id ) = @_;
+
+         matrix_invite_group_users( $creator, $group_id, $user );
+      })->then( sub {
+         matrix_accept_group_invite( $group_id, $user );
+      })->then( sub {
+         matrix_get_joined_groups( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_list( my $group_ids = $body->{groups} );
+         assert_deeply_eq( $group_ids, [ $group_id ] );
+
+         matrix_remove_group_users( $creator, $group_id, $user );
+      })->then( sub {
+         matrix_get_joined_groups( $user );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_list( my $group_ids = $body->{groups} );
+         assert_deeply_eq( $group_ids, [] );
+
+         Future->done( 1 );
+      });
+   };
+
+
+push our @EXPORT, qw( matrix_invite_group_users matrix_accept_group_invite matrix_get_joined_groups matrix_remove_group_self );
+
+
+=head2 matrix_invite_group_users
+
+Invite user to group
```

Comment:
these are getting a bit shit

Reply:
I'm not really sure what else to say about them

Reply:
Ideally I'd like to see some comments on what the types of the args are (strings? objects?), and what it returns. yup it's kinda trivial but I shouldn't have to reverse-engineer it by looking at the code.

anyway, this will do for now.